contracts/OrderBookWithFee.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.11;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+
+import "./interfaces/IOrderNotificationReceiver.sol";
+import "./OrderBook.sol";
+
+/// @title Public order book that accepts fees
+contract OrderBookWithFee is OrderBook, Ownable {
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    /// @notice Maximum fee is 10 bps
+    uint256 public constant MAX_FEE = 1_000;
+    /// @notice Denominator for fee
+    uint256 public constant PCT_DENOMINATOR = 1_000_000;
+
+    /// @notice Fee for broadcasting an order. In units of PCT_DENOMINATOR
+    uint256 public fee;
+
+    /// @notice Fee recipient
+    address public feeRecipient;
+
+    event FeeChanged(uint256 oldFee, uint256 newFee);
+    event FeeRecipientChanged(address oldFeeRecipient, address newFeeRecipient);
+
+    constructor(
+        LimitOrderProtocol _limitOrderProtocol,
+        uint256 _fee,
+        address _feeRecipient
+    ) OrderBook(_limitOrderProtocol) {
+        require(_fee <= MAX_FEE, "UOB: Fee exceeds MAX_FEE");
+        fee = _fee;
+        feeRecipient = _feeRecipient;
+    }
+
+    /// @notice Admin function to change the fee rate
+    /// @param _fee The new fee
+    function changeFee(uint256 _fee) external onlyOwner {
+        require(_fee <= MAX_FEE, "UOB: Fee exceeds MAX_FEE");
+        emit FeeChanged(fee, _fee);
+        fee = _fee;
+    }
+
+    /// @notice Admin function to change the fee recipient
+    /// @param _feeRecipient The new fee recipient
+    function changeFeeRecipient(address _feeRecipient) external onlyOwner {
+        emit FeeRecipientChanged(feeRecipient, _feeRecipient);
+        feeRecipient = _feeRecipient;
+    }
+
+    /// @notice Logs an order/signature pair. Also executes the order placement fee.
+    /// @param _order The order to broadcast
+    /// @param _signature A valid signature of _order
+    /// @param _notificationTarget A notification target. Disable by providing the 0 address
+    function broadcastOrder(
+        LimitOrderProtocol.Order memory _order,
+        bytes calldata _signature,
+        address _notificationTarget
+    ) public {
+        if (feeRecipient != address(0) && fee > 0) {
+            uint256 feeAmount = _order.makingAmount.mul(fee).div(
+                PCT_DENOMINATOR
+            );
+            if (feeAmount > 0) {
+                IERC20(_order.makerAsset).safeTransferFrom(
+                    msg.sender,
+                    feeRecipient,
+                    feeAmount
+                );
+            }
+
+            if (_notificationTarget != address(0)) {
```

Comment:
this block should be moved outside of the `fee`

order notification is orthogonal to taking the fee (notifier can also read back into the orderbook to see if `fee` and `feeRecipient` is set